llvm/include/llvm/CodeGen/RegAllocPriorityAdvisor.h

```diff
@@ -56,20 +58,80 @@ class DefaultPriorityAdvisor : public RegAllocPriorityAdvisor {
   unsigned getPriority(const LiveInterval &LI) const override;
 };
 
-class RegAllocPriorityAdvisorAnalysis : public ImmutablePass {
+/// Common provider for getting the priority advisor and logging rewards.
+/// Legacy analysis forwards all calls to this provider.
+/// New analysis serves the provider as the analysis result.
+/// Expensive setup is done in the constructor, so that the advisor can be
+/// created quickly for every machine function.
+/// TODO: Remove once legacy PM support is dropped.
+class RegAllocPriorityAdvisorProvider {
 public:
   enum class AdvisorMode : int { Default, Release, Development };
 
-  RegAllocPriorityAdvisorAnalysis(AdvisorMode Mode)
-      : ImmutablePass(ID), Mode(Mode){};
-  static char ID;
+  RegAllocPriorityAdvisorProvider(AdvisorMode Mode) : Mode(Mode) {}
+
+  virtual ~RegAllocPriorityAdvisorProvider() = default;
+
+  virtual void logRewardIfNeeded(const MachineFunction &MF,
+                                 llvm::function_ref<float()> GetReward) {};
 
-  /// Get an advisor for the given context (i.e. machine function, etc)
   virtual std::unique_ptr<RegAllocPriorityAdvisor>
   getAdvisor(const MachineFunction &MF, const RAGreedy &RA) = 0;
+
+  void setAnalyses(SlotIndexes *SI) { this->SI = SI; }
```

Comment:
Why can't this just be set at construction time?

Reply:
This provider is initialized only once for the pipeline and we just set the new analysis results while running Greedy.

Reply:
OK, so this is per-function state stored in a module analysis? That doesn't sound OK. The current RegAllocPriorityAdvisorAnalysis is an ImmutablePass, and has none of the function state. getAdvisor produces the per-function instance for the run of the pass. I think the signature of getAdvisor needs to gain parameters for all the function analyses passed through during the greedy run

Reply:
Okay, removing setAnalyses would make it simpler as well.